src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt

```diff
@@ -67,14 +69,153 @@ internal class ModelCheckingStrategy(
                 continue
             }
             usedInvocations++
-            checkResult(invocationResult)?.let { return it }
+            checkResult(invocationResult)?.let { failure ->
+                if (replay && failure.trace != null) {
+                    val trace = extractDebugTrace(failure, failure.trace)
+                    testFailed(
+                        failureType = failure.type,
+                        trace = trace,
+                        version = lincheckVersion,
+                        minimalPluginVersion = MINIMAL_PLUGIN_VERSION
+                    )
+
+                    doReplay()
+                    while (shouldReplayInterleaving()) {
+                        doReplay()
+                    }
+                }
+                return failure
+            }
             // get new unexplored interleaving
             currentInterleaving = root.nextInterleaving() ?: break
         }
         return null
     }
 
+    override fun shouldInvokeBeforeEvent(): Boolean {
+        if (!replay) return false
+        if (!collectTrace) return false
+        val thread = (Thread.currentThread() as? TestThread) ?: return false
+
+        return thread.inIgnoredSection && suddenInvocationResult == null
+    }
+
+
+    private val LincheckFailure.type: String
```

Comment:
Why do you need it? How does the plugin use this information? Please clarify in the code.